Modified batched residuals return of torch.linalg.lstsq

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54725 Added cuBLAS path for torch.linalg.lstsq
* #54724 Replaced _lstsq_helper with internal dispatch
* #54723 Updated linalg.lstsq with NumPy compatible kwarg rcond
* **#54722 Modified batched residuals return of torch.linalg.lstsq**

SciPy and NumPy operate only on non-batched input and return an empty array with shape (0,) if rank(a) != n.
The behavior for non-batched inputs is NumPy and SciPy compatible and the same result is computed.
For batched inputs, if any matrix in the batch has a rank less than `n`, then an empty tensor is returned.

Differential Revision: [D27993736](https://our.internmc.facebook.com/intern/diff/D27993736)

IvanYashchuk committed:
Modified batched residuals return of torch.linalg.lstsq

SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -2805,17 +2805,36 @@ static void linalg_lstsq_out_info(
+      // if at::any(rank != n) then the residuals for i-th batch with rank[i] != n are not available
+      // and the computed result is not meaningful, we fill it with INFINITY as we cannot resize it to (0,) shape
+      if (driver != "gels") {
+        residuals.masked_fill_((rank != n).unsqueeze_(-1), INFINITY);
```
When m < n and rank == m there are infinite number of solution satisfying b - Ax = 0. Probably residuals should be filled with zero here and not infinity.

mruberry replied:
@Lezcano @nikitaved, would you like to weigh-in on this UX?

nikitaved replied:
Unique or not unique solution, in this case the residual is expected to be very close to zero.

IvanYashchuk replied:
Residual is available as a byproduct of the LAPACK calls after some post-processing (abs and pow are fast element-wise operations). Raw residual is available if gels driver is used for any input and only for rank==n inputs if gelss, gelsd drivers are used. So for other cases calculating the residual would require matrix-matrix multiplication `ax` and then calculating the norm, I think it's an unnecessarily expensive procedure, and it's not immediately clear we should do that.

IvanYashchuk replied:
I'd prefer NumPy and SciPy did not have this return at all 😔

Lezcano replied:
I may have not understood this, but this is under the `m > n` condition, right?. In that case, there isn't in general a solution, and you're always going to have some residuals. I agree with @nikitaved that it would be reasonable to return ``||ax-b||^2_F``. It is true that it performs a matrix-matrix multiplication, but I would assume that that'd be alright when compared with how expensive this operation already is (it performs an SVD).

On the other hand, for matrices that are not full-rank, this function is not differentiable, and since this is under ``if (driver != "gels")``, it means that this will be hit just by CPU users. If a user is using CPU and tries to differentiate this, it will break. As such, what we return here only matters for people that use CPU in a non-differentiable setting. In this setting a matrix-matrix multiplication is quite costly on its own...

I'd often prefer to lean towards correctness, but I completely agree that this is a return that's not particularly useful in general and it'd be best removed.

nikitaved replied:
Not returning residuals is what I actually proposed while working on `linalg.lstsq` :)

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -482,7 +482,8 @@
         - **residuals** (*Tensor*):  if :math:`m > n` then for full rank matrices in :attr:`input` the tensor encodes
             the squared residuals of the solutions, that is :math:`||\text{input} @ x - b||_F^2`.
-            If :math:`m \le n`, an empty tensor is returned instead.
+            If :math:`m \le n` and rank is not equal to :math:`n`, an empty tensor is returned instead.
```
The rank of what?

nikitaved replied:
I think it breaks the LAPACK interface. LAPACK returns residuals which only make sense under the full-rank assumption.  Shall we keep the LAPACK behavior or build on top of it?

If the full-rank assumption is violated, we could directly return `||ax - b||^2` instead of empty tensors/infinites...

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -482,7 +482,8 @@
         - **residuals** (*Tensor*):  if :math:`m > n` then for full rank matrices in :attr:`input` the tensor encodes
             the squared residuals of the solutions, that is :math:`||\text{input} @ x - b||_F^2`.
-            If :math:`m \le n`, an empty tensor is returned instead.
+            If :math:`m \le n` and rank is not equal to :math:`n`, an empty tensor is returned instead.
```
"If attr:\`input\` is a matrix and..."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -482,7 +482,8 @@
             the squared residuals of the solutions, that is :math:`||\text{input} @ x - b||_F^2`.
-            If :math:`m \le n`, an empty tensor is returned instead.
+            If :math:`m \le n` and rank is not equal to :math:`n`, an empty tensor is returned instead.
+            For the batched :attr:`input` if rank is less than :math:`n`, then a tensor filled with ``inf`` values is returned.
```
"If attr:\`input\` is a batch of matrices and..."

Is the tensor filled with infs everywhere even if this property is true for only one matrix pairing?

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -2804,17 +2804,36 @@ static void linalg_lstsq_out_info(
   // now the actual call that computes the result in-place (apply_lstsq)
   at::_lstsq_helper_(solution, rank, singular_values, infos, input_working_copy, rcond, driver);
 
+  // residuals are available only if m > n and drivers other than gelsy used
```
@nikitaved Would you like to review these changes, too?

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -2804,17 +2804,36 @@ static void linalg_lstsq_out_info(
-      raw_residuals.pow_(2);
-    } else {
-      raw_residuals.abs_().pow_(2);
+    // if the driver is gelss or gelsd then the residuals are available only if rank == n
```
This comment seems disconnected from the code around it

IvanYashchuk replied:
I guess it would be more clear to have `if (driver == "gelss" || driver == "gelsd")` instead of `if (driver != "gels")`.

mruberry reviewed (commented):
Since this is editing the lstsq docs, can this fix the linebreak in the signature, too? See https://pytorch.org/docs/master/linalg.html?highlight=lstsq#torch.linalg.lstsq.

nikitaved commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -2804,17 +2804,36 @@ static void linalg_lstsq_out_info(
+    if (compute_residuals) {
+      // LAPACK stores residuals data for postprocessing in rows n:(m-n)
+      auto raw_residuals = solution.narrow(/*dim=*/-2, /*start=*/n, /*length*/m - n);
+      if (raw_residuals.is_complex()) {
+        // in-place abs is not supported for complex tensors
+        raw_residuals = raw_residuals.abs();
+        raw_residuals.pow_(2);
+      } else {
+        raw_residuals.abs_().pow_(2);
+      }
```
This block of code could be replaced with
```
raw_residuals.mul_(raw_residuals.conj())
```
which should work for both real and complex.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -482,7 +482,8 @@
         - **residuals** (*Tensor*):  if :math:`m > n` then for full rank matrices in :attr:`input` the tensor encodes
             the squared residuals of the solutions, that is :math:`||\text{input} @ x - b||_F^2`.
-            If :math:`m \le n`, an empty tensor is returned instead.
+            If :math:`m \le n` and rank is not equal to :math:`n`, an empty tensor is returned instead.
```
I do not understand this line. I assume that residuals only make sense when the map is not surjective. As such, wouldn't it be much clearer to write in the previous line ``:math:`m \geq n` `` and in this one simply:
```
If :math:`m < n`, an empty tensor is returned instead.
```
Also, the current docs does not say what happens when ``:math:`m \geq n` `` and the matrix is not full-rank.

IvanYashchuk replied:
It should be changed to "If m <= n **_or_** rank != n (if m > n), an empty tensor is returned".
I think it's better to avoid computing the residuals, we only do this because it's available from LAPACK for stated conditions and NumPy provides it as well.

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an
empty array with shape (0,) if rank(a) != n
but in PyTorch we can't replicate this behavior for batched inputs
instead if the corresponding rank for batched input is not equal to n
we fill residuals with inf

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry, @nikitaved, @Lezcano I modified the behavior for the batched inputs to return non-empty "residuals" only if all matrices in the batch are full rank (have rank == n). This could be updated in the future when we decide what would be a better way to handle the batched "residuals", see https://github.com/pytorch/pytorch/issues/56483.

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an empty array with shape (0,) if rank(a) != n.
The behavior for non-batched inputs is NumPy and SciPy compatible and the same result is computed.
For batched inputs, if any matrix in the batch has a rank less than `n`, then an empty tensor is returned.

[ghstack-poisoned]

mruberry commented:
@lezcano and @nikitaved, would you start by taking another look?

nikitaved reviewed (approved):
LGTM! Thank you, @IvanYashchuk .

Maybe it makes sense to really abolish the residual computation at some point...

lezcano reviewed (approved):
LGTM as well!

I concur with @nikitaved in that, in my opinion, it'd be best not to return the residuals (or the singular values for the matter). Perhaps we could look into proposing this to the API standardisation?

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an empty array with shape (0,) if rank(a) != n.
The behavior for non-batched inputs is NumPy and SciPy compatible and the same result is computed.
For batched inputs, if any matrix in the batch has a rank less than `n`, then an empty tensor is returned.

Differential Revision: [D27993736](https://our.internmc.facebook.com/intern/diff/D27993736)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an empty array with shape (0,) if rank(a) != n.
The behavior for non-batched inputs is NumPy and SciPy compatible and the same result is computed.
For batched inputs, if any matrix in the batch has a rank less than `n`, then an empty tensor is returned.

Differential Revision: [D27993736](https://our.internmc.facebook.com/intern/diff/D27993736)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an empty array with shape (0,) if rank(a) != n.
The behavior for non-batched inputs is NumPy and SciPy compatible and the same result is computed.
For batched inputs, if any matrix in the batch has a rank less than `n`, then an empty tensor is returned.

Differential Revision: [D27993736](https://our.internmc.facebook.com/intern/diff/D27993736)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Modified batched residuals return of torch.linalg.lstsq"


SciPy and NumPy operate only on non-batched input and return an empty array with shape (0,) if rank(a) != n.
The behavior for non-batched inputs is NumPy and SciPy compatible and the same result is computed.
For batched inputs, if any matrix in the batch has a rank less than `n`, then an empty tensor is returned.

Differential Revision: [D27993736](https://our.internmc.facebook.com/intern/diff/D27993736)

[ghstack-poisoned]
